src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs

```diff
@@ -0,0 +1,435 @@
+        }
+
+        private bool TryClose()
+        {
+            if (Interlocked.Exchange(ref _isClosed, 1) == 0)
+            {
+                return true;
+            }
+
+            // TODO make this actually close the Http3Stream by telling msquic to close the stream.
+            return false;
+        }
+
+        private async Task HandleEncodingTask()
+        {
+            var encoder = new EncoderStreamReader(10000); // TODO get value from limits
+            while (_isClosed == 0)
+            {
+                var result = await Input.ReadAsync();
+                var readableBuffer = result.Buffer;
+                if (!readableBuffer.IsEmpty)
+                {
+                    // This should always read all bytes in the input no matter what.
+                    encoder.Read(readableBuffer);
+                }
+                Input.AdvanceTo(readableBuffer.End);
+            }
+        }
+
+        private async Task HandleDecodingTask()
+        {
+            var decoder = new DecoderStreamReader();
+            while (_isClosed == 0)
+            {
+                var result = await Input.ReadAsync();
+                var readableBuffer = result.Buffer;
+                var consumed = readableBuffer.Start;
+                var examined = readableBuffer.Start;
+                if (!readableBuffer.IsEmpty)
+                {
+                    decoder.Read(readableBuffer);
+                }
+                Input.AdvanceTo(readableBuffer.End);
+            }
+        }
+
+        private async Task SendStreamIdAsync(int id)
+        {
+            await _frameWriter.WriteStreamIdAsync(id);
+        }
+
+        private async Task SendSettingsFrameAsync()
+        {
+            await _frameWriter.WriteSettingsAsync(null);
+        }
+
+        private async Task<Http3ControlStream> CreateNewControlStreamAsync()
+        {
+            var connectionContext = await _http3Connection.Context.ConnectionFeatures.Get<IQuicCreateStreamFeature>().StartUnidirectionalStreamAsync();
+            var httpConnectionContext = new HttpConnectionContext
+            {
+                ConnectionId = connectionContext.ConnectionId,
+                ConnectionContext = connectionContext,
+                Protocols = _context.Protocols,
+                ServiceContext = _context.ServiceContext,
+                ConnectionFeatures = connectionContext.Features,
+                MemoryPool = _context.MemoryPool,
+                Transport = connectionContext.Transport,
+                TimeoutControl = _context.TimeoutControl,
+                LocalEndPoint = connectionContext.LocalEndPoint as IPEndPoint,
+                RemoteEndPoint = connectionContext.RemoteEndPoint as IPEndPoint
+            };
+
+            // TODO think about whether we need a new stream or not here.
+            return new Http3ControlStream(_http3Connection, httpConnectionContext);
+        }
+
+        private async Task<long> TryReadStreamIdAsync()
```

Comment:
Lots of little async state machines aren't great. Better to have a big one that lives for the life of the request. Also all of these internal methods should be ValueTask instead of Task.

Reply:
> Better to have a big one that lives for the life of the request.

Clarify this with an example. I don't really get what you are referring to.